src/wp-admin/includes/update.php

```diff
@@ -582,42 +580,48 @@ function wp_plugin_update_row( $file, $plugin_data ) {
-		 *                            of possible values.
-		 * @param object $response {
-		 *     An object of metadata about the available plugin update.
-		 *
-		 *     @type string   $id           Plugin ID, e.g. `w.org/plugins/[plugin-name]`.
-		 *     @type string   $slug         Plugin slug.
-		 *     @type string   $plugin       Plugin basename.
-		 *     @type string   $new_version  New plugin version.
-		 *     @type string   $url          Plugin URL.
-		 *     @type string   $package      Plugin update package URL.
-		 *     @type string[] $icons        An array of plugin icon URLs.
-		 *     @type string[] $banners      An array of plugin banner URLs.
-		 *     @type string[] $banners_rtl  An array of plugin RTL banner URLs.
-		 *     @type string   $requires     The version of WordPress which the plugin requires.
-		 *     @type string   $tested       The version of WordPress the plugin is tested against.
-		 *     @type string   $requires_php The version of PHP which the plugin requires.
-		 * }
-		 */
-		do_action( "in_plugin_update_message-{$file}", $plugin_data, $response ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
-
-		echo '</p></div></td></tr>';
+		wp_admin_notice(
+			/**
+			 * Fires at the end of the update message container in each
+			 * row of the plugins list table.
+			 *
+			 * The dynamic portion of the hook name, `$file`, refers to the path
+			 * of the plugin's primary file relative to the plugins directory.
+			 *
+			 * @since 2.8.0
+			 *
+			 * @param array  $plugin_data An array of plugin metadata. See get_plugin_data()
+			 *                            and the {@see 'plugin_row_meta'} filter for the list
+			 *                            of possible values.
+			 * @param object $response {
+			 *     An object of metadata about the available plugin update.
+			 *
+			 *     @type string   $id           Plugin ID, e.g. `w.org/plugins/[plugin-name]`.
+			 *     @type string   $slug         Plugin slug.
+			 *     @type string   $plugin       Plugin basename.
+			 *     @type string   $new_version  New plugin version.
+			 *     @type string   $url          Plugin URL.
+			 *     @type string   $package      Plugin update package URL.
+			 *     @type string[] $icons        An array of plugin icon URLs.
+			 *     @type string[] $banners      An array of plugin banner URLs.
+			 *     @type string[] $banners_rtl  An array of plugin RTL banner URLs.
+			 *     @type string   $requires     The version of WordPress which the plugin requires.
+			 *     @type string   $tested       The version of WordPress the plugin is tested against.
+			 *     @type string   $requires_php The version of PHP which the plugin requires.
+			 * }
+			 */
+			$compatibility_notice . do_action( "in_plugin_update_message-{$file}", $plugin_data, $response ), // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores,
```

Comment:
Since `do_action()` returns `void`, this appends `null` (or `''`) to the notice. While this doesn't present an issue for the message, I wonder if we should just call `do_action()` above the `wp_admin_notice()` call?